Remove unwanted exceptions from @inlinable checking

jrose-apple opened the pull request:
- Treat protocol requirements just like associated types in how they inherit usable-from-inline-ness, rather than special-casing them in the check for inlinable code.

- Enum elements are already treated this way, so we can remove a redundant check for that.

- Finally, start enforcing that `dynamic` declarations need to be `@usableFromInline` to be used in inlinable functions...in Swift 5 mode or resilient code. I didn't even add a warning in Swift 4/4.2 because it was illegal to use `@usableFromInline` on a `dynamic` declaration in Swift 4.2. (Oops.)

See also #19548, which would reasonably allow us to add that warning.

jrose-apple commented on lib/Sema/ResilienceDiagnostics.cpp:
```diff
@@ -111,21 +111,6 @@ bool TypeChecker::diagnoseInlinableDeclRef(SourceLoc loc,
-  // Dynamic declarations are not versioned because there's no
-  // global entry point.
-  if (D->isDynamic())
-    return false;
```
@slavapestov I really don't know what to do about this. Fixing this is source-breaking, but 4.2 rejects `@usableFromInline` on `dynamic` declarations.

slavapestov replied:
Does @usableFromInline matter for dynamic? There’s no public entry point.

jrose-apple replied:
`-swift-version 4.2` has shipped. We can't do changes there. I guess we'll have to limit it to 5 mode.

We definitely can't print referenced internal dynamic methods; they may be using internal types. Actually, that's just a hole in general without this.

jrose-apple committed:
Allow '@usableFromInline' on 'dynamic' decls

It actually /does/ make sense to enforce the usable-from-inline rules
on dynamic declarations, but that would break source compatibility at
this point. Just allow '@usableFromInline' to be written on 'dynamic'
declarations, so that it'll be compatible with Swift 5.

jrose-apple committed:
Remove unwanted exceptions from @inlinable checking

- Treat protocol requirements just like associated types in how they
  inherit usable-from-inline-ness, rather than special-casing them in
  the check for inlinable code.

- Enum elements are already treated this way, so we can remove a
  redundant check for that.

- Finally, start enforcing that 'dynamic' declarations need to be
  '@usableFromInline' to be used in inlinable functions...in Swift 5
  mode or resilient code. I didn't even add a warning in Swift 4/4.2
  because it was illegal to use '@usableFromInline' on a 'dynamic'
  declaration in Swift 4.2. (Oops.)

jrose-apple commented:
Okay, ready for review.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
\*sigh\* I always forget the `-enable-objc-interop`.

jrose-apple commented:
Source compat suite: [Debug](
https://ci.swift.org/job/swift-PR-source-compat-suite-debug/148/), [Release](
https://ci.swift.org/job/swift-PR-source-compat-suite/1919/)

@swift-ci Please test

jrose-apple commented:
Oops, I included the test case from #19440.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@slavapestov Ping

jrose-apple commented:
Thanks!

Just to be sure…

@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test Linux

jrose-apple commented:
Hm, actually, I guess the ABI stability checking change is "correct" here, in that we weren't recording these things properly and now we are. @nkcsgexi, @airspeedswift, do you want me to add these to the expected list, or is it better to regenerate the baseline?

airspeedswift commented:
I'm good with either. Personally, I think we're going to want to regenerate the baseline at intervals for the next short while anyway, so it'll probably only be temporary if you put them in the expected list.

nkcsgexi commented:
I agree with Ben here. Very likely we will regenerate baseline after other fixes soon, so putting them in the expected list temporarily is fine. When parsable interface is ready, we should checking parsable interface as baseline so we never need to update it.
